core/src/main/scala/org/broadinstitute/dsde/rawls/serviceperimeter/ServicePerimeterService.scala

```diff
@@ -0,0 +1,91 @@
+    dataSource.inTransaction { dataAccess =>
+      dataAccess.workspaceQuery.getWorkspacesInPerimeter(servicePerimeterName)
+    }
+  }
+
+  /**
+    * Some Service Perimeters may required that they have some additional non-Terra Google Projects that need to be in
+    * the perimeter for some other reason.  These are provided to us by the Service Perimeter stakeholders and we add
+    * them to the Rawls Config so that whenever we update the list of projects for a perimeter, these projects are
+    * always included.
+    *
+    * @param servicePerimeterName
+    * @return
+    */
+  private def loadStaticProjectsForPerimeter(servicePerimeterName: ServicePerimeterName): Seq[GoogleProjectNumber] = {
+    config.staticProjectsInPerimeters.getOrElse(servicePerimeterName, Seq.empty)
+  }
+
+  /**
+    * Takes the the name of a Service Perimeter as the only parameter.  Since multiple Billing Projects can specify the
+    * same Service Perimeter, we will:
+    * 1. Load all the Billing Projects that also use this servicePerimeterName
+    * 2. Load all the Workspaces in all of those Billing Projects
+    * 3. Collect all of the GoogleProjectNumbers from those Workspaces
+    * 4. Post that list to Google to overwrite the Service Perimeter's list of included Google Projects
+    * 5. Poll until Google Operation to update the Service Perimeter gets to some terminal state
+    * Throw exceptions if any of this goes awry
+    *
+    * @param servicePerimeterName
+    * @return Future[Unit] indicating whether we succeeded to update the Service Perimeter
+    */
+
+  def overwriteGoogleProjectsInPerimeter(servicePerimeterName: ServicePerimeterName): Future[Unit] = {
+    for {
+      workspacesInPerimeter <- collectWorkspacesInPerimeter(servicePerimeterName)
+      projectNumbers = workspacesInPerimeter.flatMap(_.googleProjectNumber) ++ loadStaticProjectsForPerimeter(servicePerimeterName)
+      projectNumberStrings = projectNumbers.map(_.value).toSet
+      operation <- gcsDAO.accessContextManagerDAO.overwriteProjectsInServicePerimeter(servicePerimeterName, projectNumberStrings)
+      result <- retryUntilSuccessOrTimeout(failureLogMessage = s"Google Operation to update Service Perimeter: ${servicePerimeterName} was not successful")(config.pollInterval, config.pollTimeout) { () =>
+        gcsDAO.pollOperation(OperationId(GoogleApiTypes.AccessContextManagerApi, operation.getName)).map {
+          case OperationStatus(false, _) => Future.failed(new RawlsException(s"Google Operation to update Service Perimeter ${servicePerimeterName} is still in progress..."))
+          // TODO: If the operation to update the Service Perimeter failed, we need to consider the possibility that
+          // the list of Projects in the Perimeter may have been wiped or somehow modified in an undesirable way.  If
+          // this happened, it would be possible for Projects intended to be in the Perimeter are NOT in that
+          // Perimeter anymore, which is a problem.
```

Comment:
I know this comment was here before, but any thoughts on what we could do?  I can think of a couple of things:

- When this exception is thrown, email the Cloud Integration team
- Write a Stackdriver monitor or something that watches for this exception and alerts us
- Diff the projects in the perimeter at the start of the method to what is in the perimeter now and log/alert that information

I don't know, these all might be terrible ideas.  Maybe others aren't as worried about it as I am.  My understanding is that if this does get messed up it's very bad.  The condition may be very unlikely and we might not know a situation that can cause it.  

I guess if Doug or anyone else can justify the risk, I would like to put that justification into the comment.  Something that asserts, "Yes, we know there's a potential hole/risk here.  Here's our justification for why we're not doing more to prevent it or sound alarms if it should occur."

Of course, we don't even know all the various things that could go wrong with the Operation that gets us to this line of code, many of which we might not be concerned about.

Reply:
I like the idea of a Stackdriver monitor, that seems like a simple and effective way to notify us.

I also wonder if we should lock down access to the projects if the perimeter is messed up? It'd be super disruptive of course, especially for projects that remained in the perimeter, so I suppose it depends on how strongly we need to try and prevent exfiltration

Reply:
I don't think an error in the API call means the perimeter is corrupted. Let's start with an alert.

Reply:
wrote a ticket for a stackdriver alert https://broadworkbench.atlassian.net/browse/CA-1210